sdks/java/core/src/main/java/org/apache/beam/sdk/io/WholeFileIO.java

```diff
@@ -0,0 +1,241 @@
+                  try {
+                    ReadableByteChannel channel = FileSystems.open(resourceId);
+                    ByteBuffer byteBuffer = ByteBuffer.allocate(8192);
+                    ByteString byteString = ByteString.EMPTY;
+
+                    while (channel.read(byteBuffer) != -1) {
+                      byteBuffer.flip();
+                      byteString = byteString.concat(ByteString.copyFrom(byteBuffer));
+                      byteBuffer.clear();
+                    }
+
+                    KV<String, byte[]> kv = KV.of(
+                        resourceId.getFilename(),
+                        byteString.toByteArray()
+                    );
+
+                    c.output(kv);
+                  } catch (IOException e) {
+                    e.printStackTrace();
+                  }
+                }
+              }
+          )
+      );
+
+      return files;
+    }
+  }
+
+  /**
+   * Implements write().
+   */
+  @AutoValue
+  public abstract static class Write extends PTransform<PCollection<KV<String, byte[]>>, PDone> {
+
+    @Nullable abstract ValueProvider<ResourceId> getOutputDir();
+
+    abstract Builder toBuilder();
+
+    @AutoValue.Builder
+    abstract static class Builder {
+      abstract Builder setOutputDir(ValueProvider<ResourceId> outputDir);
+
+      abstract Write build();
+    }
+
+    public Write to(String outputDir) {
+      return to(FileBasedSink.convertToFileResourceIfPossible(outputDir));
+    }
+
+    public Write to(ResourceId outputDir) {
+      return toResource(ValueProvider.StaticValueProvider.of(outputDir));
+    }
+
+    public Write toResource(ValueProvider<ResourceId> outputDir) {
+      return toBuilder().setOutputDir(outputDir).build();
+    }
+
+    @Override
+    public PDone expand(PCollection<KV<String, byte[]>> input) {
+      checkNotNull(
+          getOutputDir(),
+          "Need to set the output directory of a WholeFileIO.Write transform."
+      );
+
+      ResourceId outputDir = getOutputDir().get();
+      if (!outputDir.isDirectory()) {
+        outputDir = outputDir.getCurrentDirectory()
+                             .resolve(outputDir.getFilename(), RESOLVE_DIRECTORY);
+      }
+      final PCollectionView<ResourceId> outputDirView = input.getPipeline()
+          .apply(Create.of(outputDir))
+          .apply(View.<ResourceId>asSingleton());
+
+      input.apply(
+          ParDo.of(
+              new DoFn<KV<String, byte[]>, Void>() {
+                @ProcessElement
+                public void processElement(ProcessContext c) {
+                  KV<String, byte[]> kv = c.element();
+
+                  ResourceId outputDir = c.sideInput(outputDirView);
+                  String filename = kv.getKey();
+                  ResourceId outputFile = outputDir.resolve(filename, RESOLVE_FILE);
+
+                  byte[] bytes = kv.getValue();
+                  try {
+                    WritableByteChannel channel = FileSystems.create(outputFile, BINARY);
```

Comment:
Same comments apply here; additionally, there still are atomicity and error handling concerns: if the bundle fails, the user will have a half-written garbage file in their output directory. The proper way to handle this is to have a temp directory and write to temp directory, and when done, rename file from temp directory to final output directory.